DirectProgramming/DPC++/StructuredGrids/1d_HeatTransfer/src/1d_HeatTransfer.cpp

```diff
@@ -60,173 +63,208 @@ void Usage(string programName) {
+  cout << "Using USM\n";
+  cout << "  Kernel runs on " << q.get_device().get_info<info::device::name>() << "\n";
+
+  // Temperatures of the current and next iteration
+  float* arr = malloc_shared<float>(num_p + 2, q);
+  float* arr_next = malloc_shared<float>(num_p + 2, q);
+
+  Initialize(arr, arr_next, num_p + 2);
+
+  // Start timer
+  dpc_common::TimeInterval time;
+
+  // for each timesteps
+  for (size_t i = 0; i < num_iter; i++) {
+    auto step =
+      [=](id<1> idx) {
+	size_t k = idx + 1;
+	if (k == num_p + 1)
+	  arr_next[k] = arr[k - 1];
+	else
+	  arr_next[k] =
+	    C * (arr[k + 1] - 2 * arr[k] + arr[k - 1]) + arr[k];
+      };
+
+    q.parallel_for(range{num_p+1}, step);
+
+    // Swap arrays for next step
+    swap(arr, arr_next);
   }
 
-  return arr;
+  // Wait for all the timesteps to complete
+  q.wait_and_throw();
+  
+  // Display time used to process all time steps
+  cout << "  Elapsed time: " << time.Elapsed() << " sec\n";
+
+  CompareResults("usm", arr, arr_CPU, num_p, C);
+
+  free(arr, q);
+  free(arr_next, q);
 }
 
-//************************************
-// Function description: calculate the results computed by the host and by the
-// device.
-//************************************
-bool CompareResults(float* device_results, float* host_results,
-                    unsigned int num_point, float C) {
-  float delta = 0.001f;
-  float difference = 0.00f;
-  double norm2 = 0;
-  bool err = false;
 
-  ofstream err_file;
-  err_file.open("error_diff.txt");
+//
+// Compute heat serially on the host
+//
+float* ComputeHeatHostSerial(float* arr, float* arr_next,
+			     float C, size_t num_p,
+			     size_t num_iter) {
+  size_t i, k;
 
-  err_file << " \t idx\theat[i]\t\theat_CPU[i] \n";
+  // Set initial condition
+  Initialize(arr, arr_next, num_p + 2);
 
-  for (unsigned int i = 0; i < num_point + 2; i++) {
-    err_file << "\n RESULT: " << i << "\t" << std::setw(12) << std::left
-             << device_results[i] << "\t" << host_results[i];
+  // Iterate over timesteps
+  for (i = 0; i < num_iter; i++) {
+    for (k = 1; k <= num_p; k++)
+      arr_next[k] =
+	C * (arr[k + 1] - 2 * arr[k] + arr[k - 1]) + arr[k];
 
-    difference = fabsf(host_results[i] - device_results[i]);
-    norm2 += difference * difference;
+    arr_next[num_p + 1] = arr[num_p];
 
-    if (difference > delta) {
-      err = true;
-      err_file << ", diff: " << difference;
-    }
+    // Swap the buffers for the next step
+    swap(arr, arr_next);
   }
 
-  return err;
+  return arr;
 }
 
+
 int main(int argc, char* argv[]) {
-  unsigned int n_point;  // The number of point in 1D space
-  unsigned int
-      n_iteration;  // The number of iteration to simulate the heat propagation
+  size_t n_point;      // The number of points in 1D space
+  size_t n_iteration;  // The number of iterations to simulate the heat propagation
 
   // Read input parameters
   try {
     n_point = stoi(argv[1]);
```

Comment:
So you need to worry about the user passing in -1 as an argument?

Reply:
I added some error checking.

Reply:
Old code did not handle -1. I added some error checking.